### PR TITLE
stable/nfs-client-provisioner: run as nobody

### PR DESCRIPTION
**Pull request by bavarianbidi**
Signed-off-by: Mario Constanti <github@constanti.de>

#### What this PR does / why we need it:

if in a k8s-cluster a psp disallow running containers as root, this
PR will run the nfs-client-provisioner as user nobody.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped

### CODE REVIEW

**Comment by verwilst** [comment]
Thanks for your contribution! Actually this helm chart is already psp-enabled, but instead creates a new psp object: https://github.com/helm/charts/blob/master/stable/nfs-client-provisioner/templates/podsecuritypolicy.yaml which allows it to run as root. Could you see how far you get when you remove the podsecuritypolicy.yaml and run as nobody? I like your method better, since it's a lot more secure, we might be able to fully drop the pspEnabled stuff !

**Comment by bavarianbidi** [comment]
> Thanks for your contribution! Actually this helm chart is already psp-enabled, but instead creates a new psp object: https://github.com/helm/charts/blob/master/stable/nfs-client-provisioner/templates/podsecuritypolicy.yaml which allows it to run as root. Could you see how far you get when you remove the podsecuritypolicy.yaml and run as nobody? I like your method better, since it's a lot more secure, we might be able to fully drop the pspEnabled stuff !

in our use-case we aren't allowed to use a psp where a process can run as root!
in my first commit, the create psp value was set to `false` (default) and the existing psp in the cluster has blocked the nfs-provisioner from start.

i updated the PR with the removal of the psp and i made a test, where i create a `pvc` and a `deployment`:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: test-nfs-pvc
spec:
  storageClassName: nfs-client
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
```

and following deployment:
```
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  name: test-nfs-pvc-deployment
spec:
  replicas: 2
  selector:
    matchLabels:
      name: client
      test: nfs
  template:
    metadata:
      labels:
        name: client
        test: nfs
    spec:
      containers:
      - name: test-nfs-pvc-deployment
        image: busybox:1.27
        command: 
          - '/bin/sh'
          - '-c'
          - 'tail -f /dev/null'
        volumeMounts:
        - name: test-nfs-pvc
          mountPath: /nfs
      securityContext:
        runAsUser: 1000
        fsGroup: 1000
      volumes:
      - name: test-nfs-pvc
        persistentVolumeClaim:
          claimName: test-nfs-pvc
```

File: stable/nfs-client-provisioner/Chart.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 apiVersion: v1
-appVersion: 3.1.0
```
Comment (verwilst) [inline stable/nfs-client-provisioner/Chart.yaml]: appVersion is still 3.1.0, you need to update version instead. :-)

**Comment by bavarianbidi** [comment]
@verwilst

**Comment by verwilst** [comment]
Forgot to click submit review earlier :-)

**Comment by bavarianbidi** [comment]
> Forgot to click submit review earlier :-)

@verwilst
i hope, everything is fine now :-)

**Comment by verwilst** [comment]
I'm still seeing appVersion: 3.2.0?

**Comment by bavarianbidi** [comment]
> I'm still seeing appVersion: 3.2.0?
> Maybe I should elaborate, nfs-client-provisioner itself is at version 3.1.0, which appVersion in the chart points to. You've versioned it 3.2.0, but that version of nfs-client-provisioner doesn't exist. :-)

@verwilst 

sorry for the circumstances.
i thought, increasing the minor-version should be ok, but you prefered the revision-number!
i didn't get that point, that you try to keep the image-version in sync with the appVersion.

hope we are fine now :wink:

**Comment by verwilst** [comment]
No, https://quay.io/repository/external_storage/nfs-client-provisioner?tag=latest&tab=tags shows the latest version to be 3.1.0. Which should be what appVersion is. You've changed it to 3.1.1 now. :-)

**Comment by bavarianbidi** [comment]
@verwilst 
ok, no i'm totally confused :smile: 
you requested the change of `appVersion` some Days ago: https://github.com/helm/charts/pull/12180#discussion_r265485767

but as i understand, this change isn't required anymore, or?!

**Comment by verwilst** [comment]
My initial comment was confusing, i'm sorry :-) I meant that the latest official release of the upstream nfs-client-provisioner is still at 3.1.0, so the appVersion doesn't need to be updated.

**Comment by bavarianbidi** [comment]
ok,

> My initial comment was confusing, i'm sorry :-) I meant that the latest official release of the upstream nfs-client-provisioner is still at 3.1.0, so the appVersion doesn't need to be updated.

thank you!
hope the PR is fine now?

**Comment by verwilst** [comment]
Thanks for your work!

**Comment by roonie007** [comment]
hey there,

first if all here is our values.yml

```yaml
nfs:
  server: "XXX.XXX.XXX.XXX"
  path: /volume1/kubernetes-nfs
storageClass:
  name: nfs
  defaultClass: true
  reclaimPolicy: Delete
  archiveOnDelete: false
```

using this modifications, PV Folders in the NFS server are no longer being deleted and we are getting this error on the `nfs-client-provisioner` logs  :

```
I0326 12:44:12.841360 1 controller.go:1158] delete "pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3": started
E0326 12:44:12.854548 1 controller.go:1181] delete "pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3": volume deletion failed: open /persistentvolumes/infra-dood-pvc-pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3/containerd: permission denied
W0326 12:44:12.854597 1 controller.go:787] Retrying syncing volume "pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3" because failures 2 < threshold 15
E0326 12:44:12.854618 1 controller.go:802] error syncing volume "pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3": open /persistentvolumes/infra-dood-pvc-pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3/containerd: permission denied
I0326 12:44:12.854646 1 event.go:221] Event(v1.ObjectReference{Kind:"PersistentVolume", Namespace:"", Name:"pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3", UID:"b29649a6-4fc4-11e9-b9ec-0050568b99b2", APIVersion:"v1", ResourceVersion:"20443966", FieldPath:""}): type: 'Warning' reason: 'VolumeFailedDelete' open /persistentvolumes/infra-dood-pvc-pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3/containerd: permission denied
```

You can try it by, simply running a `helm install --name mongo --namespace mongo stable/mongodb` and set the storageClass if it's not by default, then try to remove the PVC and you will get that error.


PS: Everything works perfectly in 1.2.3, the folders in nfs are removed successfully.

**Comment by bavarianbidi** [comment]
> hey there,
> 
> first of all here is our values.yml
> 
> ```yaml
> nfs:
>   server: "XXX.XXX.XXX.XXX"
>   path: /volume1/kubernetes-nfs
> storageClass:
>   name: nfs
>   defaultClass: true
>   reclaimPolicy: Delete
>   archiveOnDelete: false
> ```
> 
> using this modifications, PVs Folders in the NFS server are no longer being deleted and we are getting this error on the `nfs-client-provisioner` logs :
> 
> ```
> I0326 12:44:12.841360 1 controller.go:1158] delete "pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3": started
> E0326 12:44:12.854548 1 controller.go:1181] delete "pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3": volume deletion failed: open /persistentvolumes/infra-dood-pvc-pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3/containerd: permission denied
> W0326 12:44:12.854597 1 controller.go:787] Retrying syncing volume "pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3" because failures 2 < threshold 15
> E0326 12:44:12.854618 1 controller.go:802] error syncing volume "pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3": open /persistentvolumes/infra-dood-pvc-pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3/containerd: permission denied
> I0326 12:44:12.854646 1 event.go:221] Event(v1.ObjectReference{Kind:"PersistentVolume", Namespace:"", Name:"pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3", UID:"b29649a6-4fc4-11e9-b9ec-0050568b99b2", APIVersion:"v1", ResourceVersion:"20443966", FieldPath:""}): type: 'Warning' reason: 'VolumeFailedDelete' open /persistentvolumes/infra-dood-pvc-pvc-a71ea385-4fc4-11e9-9ca7-0050568b99b3/containerd: permission denied
> ```
> You can try it by, simply running a `helm install --name mongo --namespace mongo stable/mongodb` and set the storageClass if it's not by default, then try to remove the PVC and you will get that error.
> 
> PS: Everything works perfectly in 1.2.3, the folders in nfs are removed successfully.

Hi @ronnie007,

I think, setting `securityContext` in the `values.yaml` might be the best option to make it work for everyone! Will create a PR!
@verwilst what do you think about that?

i didn't run into this issue, because the existing NFS-shares we are using, are preprovisioned with `777`.
Are you using a preprovisioned NFS-share?
Which permissions are set?

Sorry for your issue!

**Comment by roonie007** [comment]
@bavarianbidi 

our NFS path "/volume1/kubernetes-nfs" has a permission "777", so let's say we did deploy a mongo helm, the nfs-client-provisioner will create a directory inside "/volume1/kubernetes-nfs" using this configuration
```yaml
securityContext:
  runAsUser: 65534
  fsGroup: 65534
```

then the mongo deployment start writing its own files with his configuration
```yaml
securityContext:
  runAsUser: 1001
  fsGroup: 1001
```

so we will have files with root permission.

when we delete the mongo PVC, the nfs-client-provisioner will try to remove the files as user "65534" since the files are written with "1001", it will fail because "65534" does not have the permission to delete those files.

I hope this helps you to better undestand the problem we are having.

**Comment by verwilst** [comment]
Ok, so i'll just revert the changes and keep it as it was before.

**Comment by Teslavolt** [comment]
I'm not sure if this is the right place to write this, but the most recent update to the chart broke my deployment(using the older version fixes it).

I had to change some settings on our NFS mount, and redeployed this helm chart.  The provisioner seemed to deploy fine, but any time I tried to use the storageclass, it would hang.  When looking at the logs for the provisioner, it seemed the provisioner had access denied when trying mkdir on the shared nfs mount.  I tried many different things in the values file, reinstalled the whole cluster, and tried running it under a service account with cluster admin privs, but it was still unusable.

Setting --version 1.2.3 when installing the this chart fixes my issues, and claims appear to work fine again.

**Pull request merged**